tests/test_twitter_ads_bookmark.py

```diff
@@ -1,179 +1,193 @@
+                second_sync_messages = [record.get('data') for record in
+                                        second_sync_records.get(
+                                            stream, {}).get('messages', [])
+                                        if record.get('action') == 'upsert']
+                first_bookmark_value = first_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)
+                second_bookmark_value = second_sync_bookmarks.get('bookmarks', {stream: None}).get(stream)
+
+
+                if expected_replication_method == self.INCREMENTAL:
+
+                    # collect information specific to incremental streams from syncs 1 & 2
+                    first_bookmark_value_utc = self.convert_state_to_utc(
+                        first_bookmark_value)
+                    second_bookmark_value_utc = self.convert_state_to_utc(
+                        second_bookmark_value)
+
+
+                    simulated_bookmark_value = self.convert_state_to_utc(new_states['bookmarks'][stream])
+
+                    # Verify the first sync sets a bookmark of the expected form
+                    self.assertIsNotNone(first_bookmark_value)
+
+                    # Verify the second sync sets a bookmark of the expected form
+                    self.assertIsNotNone(second_bookmark_value)
+
+                    # Verify the second sync bookmark is Equal to the first sync bookmark
+                    # assumes no changes to data during test
+                    self.assertEqual(second_bookmark_value, first_bookmark_value)
+
+                    # `targeting_criteria` is child stream of line_items stream which is incremental.
+                    # We are writing a separate bookmark for the child stream in which we are storing
+                    # the bookmark based on the parent's replication key.
+                    # But, we are not using any fields from the child record for it.
+                    # In addition to it, we are just using parent's primary key to fetch the child record but
+                    # we are not storing parent record any where.
+                    if stream == "targeting_criteria":
+                        # collect parent pk value for 1st sync
+                        first_sync_child_foreign_keys = set()
+                        for record in first_sync_messages:
+                            first_sync_child_foreign_keys.add(record['line_item_id'])
+
+                        # collect parent pk value for 2nd sync
+                        second_sync_child_foreign_keys = set()
+                        for record in second_sync_messages:
+                            second_sync_child_foreign_keys.add(record['line_item_id'])
+
+                        # Verify that all foreign keys in the 2nd sync are available in in 1st sync also. 
+                        self.assertTrue(second_sync_child_foreign_keys.issubset(first_sync_child_foreign_keys))
```

Comment:
I'm not sure this assertion is proving the relationship described in the comment. Can you instead find an expected list of foreign keys on the child stream that we'd expect to be repicated on sync 2 based based on the parent bookmark?

Reply:
Here we are asserting that we are getting more parent's children in the first sync than in the second sync. We also had this https://github.com/singer-io/tap-twitter-ads/blob/092e55cdc4751d7dde79e96028c5ddf3dd3118a6/tests/test_twitter_ads_bookmark.py#L170-#L173 assertion,
are you expecting anything else?

Reply:
Added assertion to verify following both scenarios,

- Verify that each child record from 1st sync whose parent record has a replication key value greater than or equal to bookmark value is replicated in 2nd sync.
- Verify all parents of the children in 2nd sync follows the bookmark.